src/intrinsics.jl

```diff
@@ -0,0 +1,357 @@
+module KernelIntrinsics
+
+import ..KernelAbstractions: Backend
+import GPUCompiler: split_kwargs, assign_args!
+
+"""
+    get_global_size()::@NamedTuple{x::Int, y::Int, z::Int}
+
+Return the number of global work-items specified.
+
+!!! note
+    Backend implementations **must** implement:
+    ```
+    @device_override get_global_size()::@NamedTuple{x::Int, y::Int, z::Int}
+    ```
+"""
+function get_global_size end
+
+"""
+    get_global_id()::@NamedTuple{x::Int, y::Int, z::Int}
+
+Returns the unique global work-item ID.
+
+!!! note
+    1-based.
+
+!!! note
+    Backend implementations **must** implement:
+    ```
+    @device_override get_global_id()::@NamedTuple{x::Int, y::Int, z::Int}
+    ```
+"""
+function get_global_id end
+
+"""
+    get_local_size()::@NamedTuple{x::Int, y::Int, z::Int}
+
+Return the number of local work-items specified.
+
+!!! note
+    Backend implementations **must** implement:
+    ```
+    @device_override get_local_size()::@NamedTuple{x::Int, y::Int, z::Int}
+    ```
+"""
+function get_local_size end
+
+"""
+    get_local_id()::@NamedTuple{x::Int, y::Int, z::Int}
+
+Returns the unique local work-item ID.
+
+!!! note
+    1-based.
+
+!!! note
+    Backend implementations **must** implement:
+    ```
+    @device_override get_local_id()::@NamedTuple{x::Int, y::Int, z::Int}
+    ```
+"""
+function get_local_id end
+
+"""
+    get_num_groups()::@NamedTuple{x::Int, y::Int, z::Int}
+
+Returns the number of groups.
+
+!!! note
+    Backend implementations **must** implement:
+    ```
+    @device_override get_num_groups()::@NamedTuple{x::Int, y::Int, z::Int}
+    ```
+"""
+function get_num_groups end
+
+"""
+    get_group_id()::@NamedTuple{x::Int, y::Int, z::Int}
+
+Returns the unique group ID.
+
+!!! note
+    1-based.
+
+!!! note
+    Backend implementations **must** implement:
+    ```
+    @device_override get_group_id()::@NamedTuple{x::Int, y::Int, z::Int}
+    ```
+"""
+function get_group_id end
+
+"""
+    localmemory(T, dims)
+
+Declare memory that is local to a workgroup.
+
+!!! note
+    Backend implementations **must** implement:
+    ```
+    @device_override localmemory(T::DataType, ::Val{Dims}) where {T, Dims}
+    ```
+    As well as the on-device functionality.
+"""
+localmemory(::Type{T}, dims) where {T} = localmemory(T, Val(dims))
+
+"""
+    barrier()
+
+After a `barrier()` call, all read and writes to global and local memory
+from each thread in the workgroup are visible in from all other threads in the
+workgroup.
+
+!!! note
+    `barrier()` must be encountered by all workitems of a work-group executing the kernel or by none at all.
+
+!!! note
+    Backend implementations **must** implement:
+    ```
+    @device_override barrier()
+    ```
+"""
+function barrier()
+    error("Group barrier used outside kernel or not captured")
+end
```

Comment:
We should specify the memory semantics that are required by the barrier implementation

Reply:
The semantics in the docs string should be more specific

Reply:
In particular:
- https://github.com/JuliaGPU/KernelAbstractions.jl/pull/618
- https://github.com/JuliaGPU/OpenCL.jl/pull/338
- https://github.com/JuliaGPU/AMDGPU.jl/pull/783
- https://github.com/JuliaGPU/Metal.jl/issues/608